packages/replay-next/src/suspense/createGenericRangeCache.ts

```diff
@@ -0,0 +1,227 @@
+import { ExecutionPoint, PointRange } from "@replayio/protocol";
+import sortBy from "lodash/sortBy";
+
+import { ReplayClientInterface } from "shared/client/types";
+
+import {
+  Range,
+  getMissingRanges,
+  isInRange,
+  mergeRanges,
+  rangeContains,
+  rangesOverlap,
+} from "../utils/range-list";
+import { Subscriber, UnsubscribeFunction } from "./SearchCache";
+import { Thennable } from "./types";
+
+export interface RangeCache<T> {
+  getValuesSuspense(client: ReplayClientInterface, range: PointRange): T[];
+  getValuesAsync(client: ReplayClientInterface, range: PointRange): Thennable<T[]> | T[];
+  getCachedValues(range: PointRange): T[];
+}
+
+interface RangeStore<T> {
+  cachedRanges: Range[];
+  valuesFromCachedRanges: T[];
+  pendingRanges: Range[];
+  valuesFromPendingRanges: T[];
+  subscribe(subscriber: Subscriber): UnsubscribeFunction;
+  notifySubscribers(): void;
+  fetchMissingRanges(client: ReplayClientInterface, range: PointRange): Promise<void>;
+  getExecutionPoint: (value: T) => ExecutionPoint;
+}
+
+interface PendingRange<T> {
+  range: Range;
+  values: T[];
+  promise?: Promise<void>;
+}
+
+interface FailedRange {
+  range: Range;
+  error: any;
+}
+
+export function createGenericRangeCache<T>(
+  fetchRange: (
+    client: ReplayClientInterface,
+    range: PointRange,
+    cacheValues: (values: T[]) => void,
+    cacheError: (error: any) => void
+  ) => Promise<void>,
+  getExecutionPoint: (value: T) => ExecutionPoint
+): RangeCache<T> {
+  const pendingRanges: PendingRange<T>[] = [];
+  const failedRanges: FailedRange[] = [];
+  const subscribers: Set<Subscriber> = new Set();
+  const store: RangeStore<T> = {
+    cachedRanges: [],
+    valuesFromCachedRanges: [],
+    pendingRanges: [],
+    valuesFromPendingRanges: [],
+    subscribe(subscriber: Subscriber) {
+      subscribers.add(subscriber);
+      return () => {
+        subscribers.delete(subscriber);
+      };
+    },
+    notifySubscribers() {
+      subscribers.forEach(subscriber => subscriber());
+    },
+    async fetchMissingRanges(client: ReplayClientInterface, pointRange: PointRange) {
+      const requestedRange = pointRangeToRange(pointRange);
+      const missingRanges = getMissingRanges(
+        requestedRange,
+        mergeRanges(store.cachedRanges, store.pendingRanges)
+      );
+
+      const failedRange = failedRanges.find(failedRange =>
+        missingRanges.some(missingRange => rangeContains(missingRange, failedRange.range))
+      );
+      if (failedRange) {
+        throw failedRange.error;
+      }
```

Comment:
Won't this potentially give us false positives?

If we try to fetch all events for a whole recording, and it fails, then wouldn't any subsequent focus ranges be flagged as failed?

Reply:
No, it's the other way around: only ranges that completely contain a failed range will be flagged as failed. (Yeah, I find contains type functions confusing as well)